### DISCUSSION TITLE
USWDS 2.13.0

### DISCUSSION THREAD

**Discussion by thisisdano** [uswds/uswds#4431] [General]
# What's new in USWDS 2.13.0

## Breaking changes
⚠️ **Note: This release contains a couple accessibility-related updates that require a manual markup change.** 

⚠️ **Improved resilience of icon-only functionality.** We updated a couple components that use icon-only buttons so that they provide a text equivalent if the image path is broken and does not load. Specifically, this applies to two specific components:
- **The `small` variant of the Search button:** The button that includes an image of a magnifying glass
- **The social icons in the Footer:** This includes the Facebook, Twitter, YouTube, Instagram, and RSS buttons. 

If you use these components, they will require a markup change.

#### Old markup
```
<button class="usa-button" type="submit">
  <span class="usa-sr-only">Search</span>
</button>
```

#### New markup as of 2.13.0
```
<button class="usa-button" type="submit">    
  <img src="{{ your uswds image path }}/usa-icons-bg/search--white.svg" class="usa-search__submit-icon" alt="Search">
</button>
```

#### Old markup
```
<a class="usa-social-link usa-social-link--facebook" href="{{ link }}">
  <span>Facebook</span>
</a>
<a class="usa-social-link usa-social-link--twitter" href="{{ link }}">
  <span>Twitter</span>
</a>
<a class="usa-social-link usa-social-link--youtube" href="{{ link }}">
  <span>YouTube</span>
</a>
<a class="usa-social-link usa-social-link--instagram" href="{{ link }}">
  <span>Instagram</span>
</a>
<a class="usa-social-link usa-social-link--rss" href="{{ link }}">
  <span>RSS</span>
</a>

```

#### New markup as of 2.13.0
```
<a class="usa-social-link" href="{{ link }}">
  <img class="usa-social-link__icon" src="{{ your uswds image path }}/usa-icons/facebook.svg" alt="Facebook">
</a>
<a class="usa-social-link" href="{{ link }}">
  <img class="usa-social-link__icon" src="{{ your uswds image path }}/usa-icons/twitter.svg" alt="Twitter">
</a>
<a class="usa-social-link" href="{{ link }}">
  <img class="usa-social-link__icon" src="{{ your uswds image path }}/usa-icons/youtube.svg" alt="YouTube">
</a>
<a class="usa-social-link" href="{{ link }}">
  <img class="usa-social-link__icon" src="{{ your uswds image path }}/usa-icons/instagram.svg" alt="Instagram">
</a>
<a class="usa-social-link" href="{{ link }}">
  <img class="usa-social-link__icon" src="{{ your uswds image path }}/usa-icons/rss_feed.svg" alt="RSS">
</a>

```

## Other improvements and bug fixes
**Fixed deprecation workings in our Sass compilation.** Refactored our code to avoid division! (https://github.com/uswds/uswds/pull/4314) And avoid instances of unquoted string interpolation. (https://github.com/uswds/uswds/pull/4380) Thanks @aduth!

**File upload confirms files to screenreaders.** Now File Input will tell screenreaders the total number of files and the names of files added to the component. (https://github.com/uswds/uswds/pull/4415)

**Mobile navigation now makes background content inert.** When the mobile navigation is active, all other non-nav content is hidden. This prevents accidentally leaving the focus of the active mobile menu. (https://github.com/uswds/uswds/pull/4411)

⚠️ **Improve screenreader experience of Date Picker.** Now screen readers can better describe the label and description of a date picker `input`. Thanks @mahoneycm! (https://github.com/uswds/uswds/pull/4414)

This one needs a manual markup change:

**Old Date Picker:** Two IDs in `aria-describedby`
```
<label class="usa-label" id="{{ label ID }}" for="{{ input ID }}">{{ label text }}</label>
<div class="usa-hint" id="{{ hint ID }}">{{ hint text }}</div>
<div class="usa-date-picker">
    <input
        class="usa-input"
        id="{{ input ID }}"
        name="{{ input ID }}"
        type="text"
        aria-describedby="{{ label ID }} {{ hint ID }}"
        >
</div>
```

**New Date Picker:** Label ID in `aria-labelledby` and hint ID in `aria-describedby`
```
<label class="usa-label" id="{{ label ID }}" for="{{ input ID }}">{{ label text }}</label>
<div class="usa-hint" id="{{ hint ID }}">{{ hint text }}</div>
<div class="usa-date-picker">
    <input
        class="usa-input"
        id="{{ input ID }}"
        name="{{ input ID }}"
        type="text"
        aria-labelledby="{{ label ID }}"
        aria-describedby="{{ hint ID }}"
        >
</div>
```

## Dependencies

| Package                           | Old     | New     |
| :-------------------------------- | :------ | :------ |
| @18f/identity-stylelint-config    | —       | 1.0.0   |
| chrome-launcher                   | 0.14.1  | 0.15    |
| eslint                            | 7.32.0  | 8.4.1   |
| eslint-config-airbnb-base         | 14.2.1  | 15.0.0  |
| eslint-plugin-no-unsanitized      | 3.2.0   | 4.0.1   |
| gulp-eslint                       | 6.0.0   | removed |
| gulp-stylelint                    | 13.0.0  | removed |
| jsdom                             | 17.0.0  | 19.0.0  |
| mocha                             | 9.1.3   | 6.2.0   |
| sinon                             | 11.1.2  | 12.0.1  |
| stylelint                         | 13.13.0 | 14.1.0  |
| stylelint-config-prettier         | 8.0.2   | removed |
| stylelint-config-recommended-scss | 4.3.0   | removed |
| stylelint-prettier                | 1.2.0   | removed |
| stylelint-scss                    | 3.21.0  | removed |

`0` vulnerabilities in regular dependencies (dependencies for USWDS projects installed with `npm install uswds`)

**Internal only:** `8 low, 17 moderate, 12 high, 1 critical` vulnerabilities in devDependencies (development dependencies)

**Release ZIP SHA-256 hash:** `34a951034e07288c6a23c2db3b9e54d6b5f11bec13b1484ca8a8ca8f3b013fbf`

<hr /><em>This discussion was created from the release <a href='https://github.com/uswds/uswds/releases/tag/v2.13.0'>USWDS 2.13.0</a>.</em>

**Comment by davemcorwin** [comment]
Are the changes to the markup only required to take advantage of the "new" functionality, or will this update cause the styles to change if the markup is untouched?

**Reply by mejiaj** [reply]
Sorry, this slipped through! If you update without the markup changes it will break.

**Comment by rogeruiz** [comment]
hey @thisisdano 👋🏽 I'm not sure if this is the best place to ask, but is it possible to get some advice on the upgrade path between `v2.11.2` to `v2.13.0`? I'm looking to get rid of these deprecation warnings in `trussworks/react-uswds` and could use this PR, but it seems like there are some structural changes to markup in the versions changes. Is that last assumption correct here?

**Reply by mejiaj** [reply]
Hey @rogeruiz! The latest release is [2.13.2](https://github.com/uswds/uswds/releases/tag/v2.13.2). 

We try our best to call out these breaking changes on each release. So checking out the release notes is your best bet at identifying breaking changes. 

Since 2.11.2 there's been markup changes to [Date range](https://github.com/uswds/uswds/pull/4270). Besides that 2.12.0 has new/updated settings. You can view those [here](https://github.com/uswds/uswds/releases/tag/v2.12.0).
